### ISSUE TITLE
PortableGtkReactor has latency of 100ms between select() calls

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5396]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5396|
|Type|defect|
|Created|2011-11-17 04:10:39Z|
|Branch|https://github.com/twisted/twisted/tree/portablegtk-5396|

Waiting 100ms between every network event (unless some `callLater` happens to shorten this) means tests with short timeouts will often fail, and in general applications will be slow and unresponsive. It is also slowing down our Windows test runs by about ~30 seconds on any machine with gtk2 installed. 

I propose checking for network events every 10ms instead. Still not great, but an improvement.

<details><summary>Searchable metadata</summary>

```
trac-id__5396 5396
type__defect defect
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_portablegtk_5396 branches/portablegtk-5396
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1321503039000000 1321503039000000
changetime__1323370937000000 1323370937000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [33188]) Branching to 'portablegtk-5396'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/portablegtk-5396

I don't think there's any point (or good way of running consistently) a unit test for this, given how broken the whole concept is. We will however see faster builds and less failed tests due to timeouts on those Windows buildslaves that have gtk installed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

And now bzr decided to break on the Windows machines. I'll rerun tests in the morning.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst|
|-|-|

This change makes the problem described in [[#4376](https://github.com/twisted/twisted/issues/4376)](https://github.com/twisted/twisted/issues/4376) far worse.  I suggest closing this as invalid or duplicate and finishing #4376 instead.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst removed owner|
|-|-|

1. This change doesn't affect the regular gtk reactor, only the portable one.

2. Given the way the portable one is implemented, I don't think the solution for [#4376](https://github.com/twisted/twisted/issues/4376) has any relevance. In fact, if we did implement it for the portable reactor as well, network events would only ever be handled if someone did a GUI action! The tests running with this reactor would simply sit there forever, and then timeout.

The other alternative to this solution is running gtk in a thread with something like threaded select reactor, or the reverse thing we did for serial ports. Even if that's the case, in the short term this solution will make network programs more responsive on windows, albeit at the expense of higher power usage, yes.

If this is rejected again my "solution" for [#5392](https://github.com/twisted/twisted/issues/5392) is going to be to increase timeouts on the tests, since that's the only way to make them pass. Pick your poison!

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

Just to be more explicit: there are two difference places with the 100ms callLater, with two completely different use cases. In the normal gtk reactor it's unnecessary ([#4376](https://github.com/twisted/twisted/issues/4376)); in the portable gtk reactor it's the only way network events are ever handled (this ticket).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @itamarst|
|-|-|

I suppose this is fine.  My _incredibly_ scientific test involving a Windows VM and one sample showed this change speeding the test suite up from 280s to 230s.  It's not actually clear to me that `PortableGtkReactor` is usable for any real applications, but that's just as true whether I look at it before or after this change.

I probably wouldn't go out on the limb of claiming this change as a "bugfix", but I'll leave that decision entirely up to you. :)

Go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst set status to `closed`|
|-|-|

(In [33296]) Merge portablegtk-5396.

Fixes: [#5396](https://github.com/twisted/twisted/issues/5396)
Author: itamar
Review: exarkun

Make portable gtk2 reactor more responsive, speeding up trial while we're at it... at the cost of greater power usage.
